File: solution/1717. Maximum Score From Removing Substrings/1717.cpp
class Solution {
public:
    int maximumGain(string s, int x, int y) {
        char a = 'a', b = 'b';
        if (x < y) { // 保證 'ab', 即 x 是最大的
            swap(a, b);
            swap(x, y);
        }
        string s1 = remove(s, a, b); // 先移除 'ab'
        string s2 = remove(s1, b, a); // 後移除 'ba'

        return (s.size() - s1.size()) / 2 * x + (s1.size() - s2.size()) / 2 * y;
    }
    
private:
    string remove(string& s, char& a, char& b) {
        string st;
        for (char& c : s) {
            if (!st.empty() && st.back() == a && c == b)
                st.pop_back();
            else
                st.push_back(c);
        }
        return st;
    }
};

// ab -> x
// ba -> y
// aba -> get(ba) -> a
//     -> get(ab) -> a
// bab -> get(ab) -> b
//     -> get(ba) -> b

// cdbcbaabab -> ba = 5
// cdbcabab -> ba = 5
// cdbcab -> ab = 4
// cdbc -> ab = 4

// cdbcbaabab -> ba = 5
// cdbcabab -> ba = 5
// cdbcab -> ba = 5
// cdbc -> ab = 4

File: solution/1717. Maximum Score From Removing Substrings/1717.py
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        a, b = 'a', 'b'
        if x < y:
            a, b = b, a
            x, y = y, x
        
        def remove(s, a, b):
            st = []
            for c in s:
                if st and st[-1] == a and c == b:
                    st.pop()
                else:
                    st.append(c)
            return ''.join(st)
        
        s1 = remove(s, a, b)
        s2 = remove(s1, b, a)
        return (len(s) - len(s1)) // 2 * x + ((len(s1) - len(s2)) // 2 * y)
        
